crates/sui-core/src/authority/authority_store.rs

```diff
@@ -354,19 +360,32 @@ impl<
     pub async fn get_transaction_envelope(
         &self,
         object_ref: &ObjectRef,
-    ) -> Result<Option<TransactionEnvelope<S>>, SuiError> {
+    ) -> SuiResult<Option<TransactionEnvelope<S>>> {
         let transaction_option = self
             .lock_service
             .get_lock(*object_ref)
             .await?
             .ok_or(SuiError::TransactionLockDoesNotExist)?;
 
+        // Returns None if either no TX with the lock, or TX present but no entry in transactions table.
+        // However we retry a couple times because the TX is written after the lock is acquired, so it might
+        // just be a race.
         match transaction_option {
-            Some(tx_digest) => Ok(Some(
-                self.transactions
-                    .get(&tx_digest)?
-                    .expect("Stored a lock without storing transaction?"),
-            )),
+            Some(tx_digest) => {
+                // Retry getting transaction
+                retry_with_index(Exponential::from_millis(20).map(jitter).take(3), |retry_num| {
+                    match self.transactions.get(&tx_digest) {
```

Comment:
Since this does not yield to the tokio executor, in case the executor is single threaded the task that actually writes to the DB may never be executed, and as a result this would simply delay returning the value. What about using an async friendly retry library here: https://crates.io/crates/tokio-retry

Reply:
Yeah I might just add that.